firestore/src/jni/arena_ref.h

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIREBASE_FIRESTORE_SRC_JNI_ARENA_REF_H_
+#define FIREBASE_FIRESTORE_SRC_JNI_ARENA_REF_H_
+
+#include <utility>
+
+#include "app/src/assert.h"
+#include "firestore/src/jni/env.h"
+#include "firestore/src/jni/object.h"
+#include "firestore/src/jni/object_arena.h"
+
+namespace firebase {
+namespace firestore {
+namespace jni {
+
+/**
+ * An RAII wrapper that uses ObjectArena to manage the reference. It
+ * automatically deletes the JNI local reference when it goes out of scope.
+ * Copies and moves are handled by creating additional references as required.
+ */
+class ArenaRef {
+ public:
+  ArenaRef() = default;
+
+  explicit ArenaRef(const Object& ob) {
+    Env env(GetEnv());
+    assignState(ObjectArena::GetInstance().Put(env, ob));
+  }
+
+  ArenaRef(const ArenaRef& other) {
+    if (other.id_.first) {
+      Env env(GetEnv());
+      assignState(ObjectArena::GetInstance().Dup(env, other.id_.second));
+    }
+  }
+
+  ArenaRef& operator=(const ArenaRef& other) {
+    if (id_ != other.id_) {
+      Env env(GetEnv());
+      if (id_.first) {
+        ObjectArena::GetInstance().Remove(env, id_.second);
+      }
+      reset();
+      if (other.id_.first) {
+        assignState(ObjectArena::GetInstance().Dup(env, other.id_.second));
+      }
+    }
+    return *this;
+  }
+
+  ArenaRef(ArenaRef&& other) noexcept : ArenaRef(other.release()) {}
```

Comment:
This is going to fail if other.id_.first is not true. Please improve the robustness.